File: desktop/src/lib/workspace.ts
import { moveBefore, moveToEnd } from "@/lib/ordering"
import {
  activateIndex,
  activateNext,
  activatePrevious,
  closeOtherTabs,
  closeTab,
  openTab,
  tabState,
  type TabState,
} from "@/lib/tabs"

/**
 * One or two panes side by side, plus which one holds focus — ported from
 * ADBKit's `Workspace`.
 *
 * The fiddly rules are all here rather than in a component, for the reason the
 * Swift version gives: collapsing an emptied pane, keeping ids globally unique,
 * keeping a pane focused, and never leaving the workspace empty are exactly the
 * things that go wrong and exactly the things worth testing without a window.
 *
 * A feature id is open in at most one pane, so moving a tab across is a move
 * and not a copy.
 */
export interface Workspace {
  /** The open panes, left to right — always 1 or 2. */
  readonly groups: readonly TabState[]
  /** Which pane new tabs, close and cycle act on. Always a valid index. */
  readonly focusedGroup: number
}

/** A fresh workspace with a single pane showing `fallback`. */
export function newWorkspace(fallback: string): Workspace {
  return { groups: [tabState([fallback], fallback)], focusedGroup: 0 }
}

/**
 * Rebuild from persisted panes, enforcing every invariant: at most two panes,
 * ids valid and globally unique, a non-empty result, and an in-range focus.
 */
export function restoreWorkspace(
  persisted: readonly { tabs: readonly string[]; activeTab: string | null }[],
  focusedGroup: number,
  fallback: string,
  isKnownTab: (id: string) => boolean,
): Workspace {
  const seen = new Set<string>()
  const groups: TabState[] = []
  for (const group of persisted.slice(0, 2)) {
    const valid = group.tabs.filter((id) => {
      if (!isKnownTab(id) || seen.has(id)) return false
      seen.add(id)
      return true
    })
    if (valid.length > 0) groups.push(tabState(valid, group.activeTab))
  }
  if (groups.length === 0) return newWorkspace(fallback)
  return { groups, focusedGroup: Math.min(Math.max(focusedGroup, 0), groups.length - 1) }
}

// MARK: reads

/** The focused pane's active tab — what the sidebar highlights. */
export function activeTab(workspace: Workspace): string | null {
  return workspace.groups[workspace.focusedGroup]?.activeTab ?? null
}

/** Every pane's active tab. Both fronts are on screen, so both are "active". */
export function activeTabs(workspace: Workspace): string[] {
  return workspace.groups.flatMap((group) => (group.activeTab === null ? [] : [group.activeTab]))
}

export function isSplit(workspace: Workspace): boolean {
  return workspace.groups.length > 1
}

export function groupIndexOf(workspace: Workspace, id: string): number {
  return workspace.groups.findIndex((group) => group.openTabs.includes(id))
}

/** Every open tab, in pane order — what the keep-alive body renders. */
export function allOpenTabs(workspace: Workspace): string[] {
  return workspace.groups.flatMap((group) => [...group.openTabs])
}

// MARK: mutations

/** Open `id`, or refocus it wherever it already is. New tabs land in focus. */
export function open(workspace: Workspace, id: string): Workspace {
  const existing = groupIndexOf(workspace, id)
  const at = existing === -1 ? workspace.focusedGroup : existing
  return { groups: replace(workspace.groups, at, (group) => openTab(group, id)), focusedGroup: at }
}

/**
 * Close `id` wherever it is. Collapses an emptied second pane; reseeds
 * `fallback` when the last pane would empty; keeps focus in range.
 */
export function close(workspace: Workspace, id: string, fallback: string): Workspace {
  const at = groupIndexOf(workspace, id)
  if (at === -1) return workspace
  const closed = closeTab(workspace.groups[at] ?? tabState([]), id)
  if (closed.openTabs.length > 0) {
    return {
      groups: replace(workspace.groups, at, () => closed),
      focusedGroup: workspace.focusedGroup,
    }
  }
  if (workspace.groups.length > 1) {
    const groups = workspace.groups.filter((_, index) => index !== at)
    return { groups, focusedGroup: Math.min(workspace.focusedGroup, groups.length - 1) }
  }
  // The last pane never goes empty.
  return { groups: [tabState([fallback], fallback)], focusedGroup: 0 }
}

/** Close every tab in `id`'s pane except it and the permanent `keep`. */
export function closeOthers(workspace: Workspace, id: string, keep: string): Workspace {
  const at = groupIndexOf(workspace, id)
  if (at === -1) return workspace
  return {
    groups: replace(workspace.groups, at, (group) => closeOtherTabs(group, id, keep)),
    focusedGroup: at,
  }
}

/**
 * Move `id` into pane `dest`, appending and activating it. Collapses the source
 * pane if it empties. A no-op for the same pane or an index that is not there.
 */
export function move(workspace: Workspace, id: string, dest: number): Workspace {
  const src = groupIndexOf(workspace, id)
  if (src === -1 || src === dest || !(dest >= 0 && dest < workspace.groups.length)) return workspace
  const groups = workspace.groups.map((group, index) => {
    if (index === src) return closeTab(group, id)
    if (index === dest) return openTab(group, id)
    return group
  })
  const collapsed = groups.filter((group, index) => index !== src || group.openTabs.length > 0)
  const landed = collapsed.findIndex((group) => group.openTabs.includes(id))
  return { groups: collapsed, focusedGroup: landed === -1 ? 0 : landed }
}

/**
 * Split: move `id` into a new second pane. Needs its pane to hold more than one
 * tab — something has to stay behind — and no existing split.
 */
export function split(workspace: Workspace, id: string): Workspace {
  const src = groupIndexOf(workspace, id)
  if (workspace.groups.length !== 1 || src === -1) return workspace
  if ((workspace.groups[src]?.openTabs.length ?? 0) < 2) return workspace
  return {
    groups: [closeTab(workspace.groups[src] ?? tabState([]), id), tabState([id], id)],
    focusedGroup: 1,
  }
}

/** Reorder `id` within its own pane so it sits before `target` (null = end). */
export function reorder(workspace: Workspace, id: string, target: string | null): Workspace {
  const at = groupIndexOf(workspace, id)
  if (at === -1) return workspace
  return {
    groups: replace(workspace.groups, at, (group) => {
      const openTabs =
        target === null ? moveToEnd(id, group.openTabs) : moveBefore(id, target, group.openTabs)
      return { openTabs, activeTab: group.activeTab }
    }),
    focusedGroup: workspace.focusedGroup,
  }
}

/** Resolve a strip drop: reorder inside a pane, or move across and position. */
export function drop(
  workspace: Workspace,
  id: string,
  dest: number,
  target: string | null,
): Workspace {
  const src = groupIndexOf(workspace, id)
  if (src === -1) return workspace
  if (src === dest) return target === null ? workspace : reorder(workspace, id, target)
  const moved = move(workspace, id, dest)
  return target === null ? moved : reorder(moved, id, target)
}

export function cycleForward(workspace: Workspace): Workspace {
  return mapFocused(workspace, activateNext)
}

export function cycleBackward(workspace: Workspace): Workspace {
  return mapFocused(workspace, activatePrevious)
}

/** Activate the tab at a 0-based index in the focused pane. */
export function activateAt(workspace: Workspace, index: number): Workspace {
  return mapFocused(workspace, (group) => activateIndex(group, index))
}

/** Give a pane focus — clicking one of its tabs, or anywhere in its body. */
export function focus(workspace: Workspace, index: number): Workspace {
  if (index === workspace.focusedGroup) return workspace
  if (!(index >= 0 && index < workspace.groups.length)) return workspace
  return { groups: workspace.groups, focusedGroup: index }
}

function mapFocused(workspace: Workspace, transform: (group: TabState) => TabState): Workspace {
  return {
    groups: replace(workspace.groups, workspace.focusedGroup, transform),
    focusedGroup: workspace.focusedGroup,
  }
}

function replace(
  groups: readonly TabState[],
  at: number,
  transform: (group: TabState) => TabState,
): TabState[] {
  return groups.map((group, index) => (index === at ? transform(group) : group))
}

File: ADBKit/Sources/ADBKit/Features/Workspace.swift
import Foundation

/// The editor-group workspace: one or two panes (`TabState` groups) side by
/// side, plus which one holds focus. Pure value type kept out of the SwiftUI
/// layer (like `TabState`/`SidebarOrdering`) so the fiddly multi-group rules —
/// collapsing an emptied pane, shifting indices, global id uniqueness, keeping
/// a pane focused, and never leaving the workspace empty — are unit-tested
/// without a UI or a device.
///
/// A feature id is open in at most one group, so moving a tab between panes is a
/// move, not a copy. `fallback` is the id seeded when the workspace would
/// otherwise be empty (the app passes "home") — kept as data so this type stays
/// free of any UI/registry knowledge.
public struct Workspace: Sendable, Equatable {
    /// The tab seeded when the last tab would close (never leave an empty pane).
    public let fallback: String
    /// The open panes, left to right — always 1 or 2.
    public private(set) var groups: [TabState]
    /// Which pane holds focus (index into `groups`) — new tabs, close, and cycle
    /// act on it. Always a valid index.
    public private(set) var focusedGroup: Int

    /// A fresh workspace with a single pane showing `fallback`.
    public init(fallback: String) {
        self.fallback = fallback
        groups = [TabState(openTabs: [fallback], activeTab: fallback)]
        focusedGroup = 0
    }

    /// Rebuild from persisted panes, enforcing every invariant: at most two
    /// panes, ids valid (`isValidID`) and globally unique, a non-empty result
    /// (seed `fallback`), and an in-range `focusedGroup`.
    public init(
        restoring persisted: [TabGroupState],
        focusedGroup: Int?,
        fallback: String,
        isValidID: (String) -> Bool
    ) {
        self.fallback = fallback
        var restored: [TabState] = []
        var seen = Set<String>()
        for group in persisted.prefix(2) {
            let valid = group.tabs.filter { isValidID($0) && seen.insert($0).inserted }
            guard !valid.isEmpty else { continue }
            restored.append(TabState(openTabs: valid, activeTab: group.activeTab))
        }
        groups = restored.isEmpty ? [TabState(openTabs: [fallback], activeTab: fallback)] : restored
        self.focusedGroup = min(max(focusedGroup ?? 0, 0), groups.count - 1)
    }

    // MARK: - Reads

    /// The focused pane's active tab — drives the device bar, sidebar highlight,
    /// and window title.
    public var activeTab: String? {
        groups.indices.contains(focusedGroup) ? groups[focusedGroup].activeTab : nil
    }
    /// Every pane's active tab (both fronts).
    public var activeTabs: Set<String> { Set(groups.compactMap(\.activeTab)) }
    /// True when split into two panes.
    public var isSplit: Bool { groups.count > 1 }
    /// Total open tabs across all panes.
    public var totalTabs: Int { groups.reduce(0) { $0 + $1.openTabs.count } }

    public func openTabs(inGroup index: Int) -> [String] {
        groups.indices.contains(index) ? groups[index].openTabs : []
    }
    public func activeTab(inGroup index: Int) -> String? {
        groups.indices.contains(index) ? groups[index].activeTab : nil
    }
    public func groupIndex(of id: String) -> Int? {
        groups.firstIndex { $0.openTabs.contains(id) }
    }

    // MARK: - Mutations

    /// Open `id`, or refocus it wherever it's already open. A not-yet-open id
    /// lands in the focused pane.
    public mutating func open(_ id: String) {
        if let group = groupIndex(of: id) {
            focusedGroup = group
            groups[group].open(id)
        } else {
            groups[focusedGroup].open(id)
        }
    }

    /// Close `id` (in whichever pane holds it). Collapses an emptied second pane;
    /// reopens `fallback` if the last pane would empty; keeps focus in range.
    public mutating func close(_ id: String) {
        guard let group = groupIndex(of: id) else { return }
        groups[group].close(id)
        guard groups[group].openTabs.isEmpty else { return }
        if groups.count > 1 {
            groups.remove(at: group)
            focusedGroup = min(focusedGroup, groups.count - 1)
        } else {
            groups[group].open(fallback) // never leave the workspace empty
        }
    }

    /// Move `id` into pane `dest` (append + activate). Collapses the source pane
    /// if it empties. No-op for the same pane or an invalid `dest`.
    public mutating func move(_ id: String, toGroup dest: Int) {
        guard let src = groupIndex(of: id), src != dest, groups.indices.contains(dest) else { return }
        groups[src].close(id)
        groups[dest].open(id)
        if groups[src].openTabs.isEmpty { groups.remove(at: src) }
        focusedGroup = groupIndex(of: id) ?? focusedGroup
    }

    /// Whether `id` can leave this workspace for *another window*.
    ///
    /// Only that it is open: moving a window's last tab into another window is
    /// a legitimate way to consolidate two windows, and leaves this one showing
    /// `fallback` exactly as closing that tab would.
    public func canDetach(_ id: String) -> Bool {
        groupIndex(of: id) != nil
    }

    /// Whether `id` can leave for a window *of its own*.
    ///
    /// Stricter, because something has to stay behind: a workspace whose only
    /// tab left for a new window would not be a move, it would be the *window*
    /// moving — and the window already moves. Same "something must stay"
    /// precondition `split` carries, counted across both panes rather than
    /// within one, because detaching collapses an emptied pane.
    public func canDetachToNewWindow(_ id: String) -> Bool {
        canDetach(id) && totalTabs > 1
    }

    /// Remove `id` from this workspace and return it, for a handoff to another
    /// window.
    ///
    /// Deliberately the same removal `close` performs — collapse an emptied
    /// second pane, keep `focusedGroup` in range, and never leave the workspace
    /// empty — because a window that has just given its last tab away should
    /// look exactly like one whose last tab was closed, rather than a blank.
    public mutating func detach(_ id: String) -> String? {
        guard canDetach(id) else { return nil }
        close(id)
        return id
    }

    /// Split the workspace: move `id` into a new second pane. Requires its pane
    /// to hold more than one tab (something must stay behind) and no existing
    /// split.
    public mutating func split(_ id: String) {
        guard groups.count == 1, let src = groupIndex(of: id), groups[src].openTabs.count > 1 else { return }
        groups[src].close(id)
        groups.append(TabState(openTabs: [id], activeTab: id))
        focusedGroup = 1
    }

    /// Reorder `id` within its own pane so it sits before `targetID` (nil = end).
    public mutating func reorder(_ id: String, before targetID: String?) {
        guard let group = groupIndex(of: id) else { return }
        let order = targetID.map { SidebarOrdering.move(id, before: $0, in: groups[group].openTabs) }
            ?? SidebarOrdering.moveToEnd(id, in: groups[group].openTabs)
        groups[group].reorder(order)
    }

    /// Resolve a strip/pane drop: reorder within the same pane, or move to `dest`
    /// and position at the drop target.
    public mutating func drop(_ id: String, intoGroup dest: Int, before targetID: String?) {
        guard let src = groupIndex(of: id) else { return }
        if src == dest {
            // A nil target is the strip's dead space past the last chip, which
            // means "put it at the end" — `reorder` already reads it that way.
            // Skipping the call for nil made that a dead drop in the tab's own
            // pane while working across panes, which is the asymmetry the strip
            // documents itself as *not* having.
            reorder(id, before: targetID)
        } else {
            move(id, toGroup: dest)
            if let targetID { reorder(id, before: targetID) }
        }
    }

    /// Activate the next / previous tab within the focused pane (wraps).
    public mutating func cycleForward() {
        guard groups.indices.contains(focusedGroup) else { return }
        groups[focusedGroup].activateNext()
    }
    public mutating func cycleBackward() {
        guard groups.indices.contains(focusedGroup) else { return }
        groups[focusedGroup].activatePrevious()
    }
    /// Activate the tab at a 0-based index in the focused pane.
    public mutating func activate(index: Int) {
        guard groups.indices.contains(focusedGroup) else { return }
        groups[focusedGroup].activate(index: index)
    }
    /// Give a pane focus (clicking one of its tabs, or its + button).
    public mutating func focus(_ index: Int) {
        if groups.indices.contains(index) { focusedGroup = index }
    }

    /// Collapse to a single pane showing `fallback` (e.g. after a role change).
    public mutating func reset() {
        groups = [TabState(openTabs: [fallback], activeTab: fallback)]
        focusedGroup = 0
    }
}
